atc/scheduler/factory/build_factory_test.go

```diff
@@ -0,0 +1,679 @@
+				"steps": [
+					{
+						"load_var": {
+							"name": "some-var",
+							"file": "some-file"
+						}
+					},
+					{
+						"load_var": {
+							"name": "some-other-var",
+							"file": "some-other-file"
+						}
+					}
+				]
+			}
+		}`,
+	},
+	{
+		Title: "in_parallel step with config",
+
+		ConfigYAML: `
+			in_parallel:
+			  steps:
+			  - load_var: some-var
+			    file: some-file
+			  - load_var: some-other-var
+			    file: some-other-file
+			  limit: 3
+			  fail_fast: true
+		`,
+
+		PlanJSON: `{
+			"in_parallel": {
+				"steps": [
+					{
+						"load_var": {
+							"name": "some-var",
+							"file": "some-file"
+						}
+					},
+					{
+						"load_var": {
+							"name": "some-other-var",
+							"file": "some-other-file"
+						}
+					}
+				],
+				"limit": 3,
+				"fail_fast": true
+			}
+		}`,
+	},
+	{
+		Title: "aggregate step",
+
+		ConfigYAML: `
+			aggregate:
+			- load_var: some-var
+			  file: some-file
+			- load_var: some-other-var
+			  file: some-other-file
+		`,
+
+		PlanJSON: `{
+			"aggregate": [
+				{
+					"load_var": {
+						"name": "some-var",
+						"file": "some-file"
+					}
+				},
+				{
+					"load_var": {
+						"name": "some-other-var",
+						"file": "some-other-file"
+					}
+				}
+			]
+		}`,
+	},
+	{
+		Title: "timeout modifier",
+
+		ConfigYAML: `
+			load_var: some-var
+			file: some-file
+			timeout: 1h
+		`,
+
+		PlanJSON: `{
+			"timeout": {
+				"step": {
+					"load_var": {
+						"name": "some-var",
+						"file": "some-file"
+					}
+				},
+				"duration": "1h"
+			}
+		}`,
+	},
+	{
+		Title: "attempts modifier",
+
+		ConfigYAML: `
+			load_var: some-var
+			file: some-file
+			attempts: 3
+		`,
+
+		CompareIDs: true,
+		PlanJSON: `{
+			"id": "4",
+			"retry": [
+				{
+					"id": "1",
+					"load_var": {
+						"name": "some-var",
+						"file": "some-file"
+					}
+				},
+				{
+					"id": "2",
+					"load_var": {
+						"name": "some-var",
+						"file": "some-file"
+					}
+				},
+				{
+					"id": "3",
+					"load_var": {
+						"name": "some-var",
+						"file": "some-file"
+					}
+				}
+			]
+		}`,
+	},
+	{
+		Title: "timeout and attempts modifier",
+
+		ConfigYAML: `
+			load_var: some-var
+			file: some-file
+			timeout: 1h
+			attempts: 3
+		`,
+
+		PlanJSON: `{
+			"retry": [
+				{
+					"timeout": {
+						"step": {
+							"load_var": {
+								"name": "some-var",
+								"file": "some-file"
+							}
+						},
+						"duration": "1h"
+					}
+				},
+				{
+					"timeout": {
+						"step": {
+							"load_var": {
+								"name": "some-var",
+								"file": "some-file"
+							}
+						},
+						"duration": "1h"
+					}
+				},
+				{
+					"timeout": {
+						"step": {
+							"load_var": {
+								"name": "some-var",
+								"file": "some-file"
+							}
+						},
+						"duration": "1h"
+					}
+				}
+			]
+		}`,
+	},
+}
+
+func init() {
+	for _, hookType := range []string{"on_success", "on_failure", "on_abort", "ensure"} {
```

Comment:
this block is generating specifications from a template - this feels like a bit of a smell to me. If all these step types have such similar behaviour that specifications can be programatically generated, might it make sense to extract a common interface that they fulfill and somehow test against that? It just seems like "multiple types with the same behaviour" is sorta the definition of polymorphism. What do you think of this observation? I'm mostly speaking from the experience of 'generating tests from a template' in this fashion in the elm codebase and finding it got a bit unwieldy and difficult to read. In the elm case, using a polymorphism technique like extensible records ended up helping.

Reply:
Yeah this part was a bit hokey. These tests are really just trying to test precedence, and with https://github.com/concourse/concourse/pull/5504 (based on this PR) precedence is determined at the parsing stage, which has its own tests, including [one big precedence test](https://github.com/concourse/concourse/pull/5504/files#diff-3ec0d2bc70f6384123558ee5a728415aR307) that uses all modifiers that shows the precedence all at once.

I can make a follow-up commit to that PR that changes these tests to just have one test per step type without having to worry about precedence.

Reply:
Added this as a TODO in the other PR so this doesn't get lost in the merge.